File: spx-knowledge-backend/app/schemas/knowledge_graph.py
"""
Knowledge Graph Schemas
知识图谱相关Schema
"""

from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field


# ============================================
# 实体相关Schema
# ============================================

class EntityBase(BaseModel):
    """实体基础Schema"""
    name: str = Field(..., description="实体名称")
    type: str = Field(..., description="实体类型")
    description: Optional[str] = Field(None, description="实体描述")
    aliases: Optional[List[str]] = Field(None, description="实体别名列表")
    confidence: Optional[float] = Field(0.7, ge=0.0, le=1.0, description="置信度分数")
    metadata: Optional[Dict[str, Any]] = Field(None, description="扩展元数据")


class EntityCreate(EntityBase):
    """创建实体Schema"""
    knowledge_base_id: int = Field(..., description="知识库ID")


class EntityUpdate(BaseModel):
    """更新实体Schema"""
    name: Optional[str] = None
    type: Optional[str] = None
    description: Optional[str] = None
    aliases: Optional[List[str]] = None
    confidence: Optional[float] = Field(None, ge=0.0, le=1.0)
    metadata: Optional[Dict[str, Any]] = None


class EntityResponse(EntityBase):
    """实体响应Schema"""
    id: int
    knowledge_base_id: int
    user_id: int
    created_at: datetime
    updated_at: datetime
    nebula_synced: bool
    nebula_synced_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


class EntityListResponse(BaseModel):
    """实体列表响应Schema"""
    id: int
    name: str
    type: str
    description: Optional[str] = None
    aliases: Optional[List[str]] = None
    confidence: float
    related_documents_count: Optional[int] = 0
    related_entities_count: Optional[int] = 0
    
    class Config:
        from_attributes = True


# ============================================
# 关系相关Schema
# ============================================

class RelationshipBase(BaseModel):
    """关系基础Schema"""
    source_entity_id: int = Field(..., description="源实体ID")
    target_entity_id: int = Field(..., description="目标实体ID")
    relation_type: str = Field(..., description="关系类型")
    description: Optional[str] = Field(None, description="关系描述")
    weight: Optional[float] = Field(0.5, ge=0.0, le=1.0, description="关系权重")
    confidence: Optional[float] = Field(0.7, ge=0.0, le=1.0, description="置信度分数")
    metadata: Optional[Dict[str, Any]] = Field(None, description="扩展元数据")


class RelationshipCreate(RelationshipBase):
    """创建关系Schema"""
    knowledge_base_id: int = Field(..., description="知识库ID")


class RelationshipUpdate(BaseModel):
    """更新关系Schema"""
    relation_type: Optional[str] = None
    description: Optional[str] = None
    weight: Optional[float] = Field(None, ge=0.0, le=1.0)
    confidence: Optional[float] = Field(None, ge=0.0, le=1.0)
    metadata: Optional[Dict[str, Any]] = None


class RelationshipResponse(RelationshipBase):
    """关系响应Schema"""
    id: int
    knowledge_base_id: int
    user_id: int
    created_at: datetime
    updated_at: datetime
    nebula_synced: bool
    nebula_synced_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


class RelationshipDetailResponse(RelationshipResponse):
    """关系详情响应Schema"""
    source_entity: Optional[EntityResponse] = None
    target_entity: Optional[EntityResponse] = None


# ============================================
# 查询相关Schema
# ============================================

class PathQueryRequest(BaseModel):
    """路径查询请求Schema"""
    source_entity_id: int = Field(..., description="源实体ID")
    target_entity_id: int = Field(..., description="目标实体ID")
    max_hops: Optional[int] = Field(3, ge=1, le=10, description="最大跳数")
    relation_types: Optional[List[str]] = Field(None, description="限制关系类型")


class PathResponse(BaseModel):
    """路径响应Schema"""
    entities: List[int] = Field(..., description="实体ID列表")
    relationships: List[Dict[str, Any]] = Field(..., description="关系列表")
    total_weight: float = Field(..., description="总权重")
    path_length: int = Field(..., description="路径长度")


class NeighborQueryRequest(BaseModel):
    """邻居查询请求Schema"""
    relation_type: Optional[str] = Field(None, description="关系类型")
    max_depth: Optional[int] = Field(1, ge=1, le=5, description="最大深度")
    limit: Optional[int] = Field(50, ge=1, le=200, description="返回数量限制")


class EntityDetailResponse(EntityResponse):
    """实体详情响应Schema"""
    relationships: List[RelationshipDetailResponse] = []
    related_documents: List[Dict[str, Any]] = []


# ============================================
# 提取相关Schema
# ============================================

class ExtractionTaskCreate(BaseModel):
    """创建提取任务Schema"""
    knowledge_base_id: int = Field(..., description="知识库ID")
    document_ids: Optional[List[int]] = Field(None, description="文档ID列表（可选，为空表示批量提取）")


class ExtractionTaskResponse(BaseModel):
    """提取任务响应Schema"""
    id: int
    knowledge_base_id: int
    document_id: Optional[int] = None
    status: str
    total_entities: int
    total_relationships: int
    error_message: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


# ============================================
# 可视化相关Schema
# ============================================

class VisualizationRequest(BaseModel):
    """可视化请求Schema"""
    knowledge_base_id: int = Field(..., description="知识库ID")
    entity_ids: Optional[List[int]] = Field(None, description="指定要展示的实体ID列表")
    relation_types: Optional[List[str]] = Field(None, description="限制关系类型")
    max_nodes: Optional[int] = Field(100, ge=1, le=1000, description="最大节点数")
    layout: Optional[str] = Field("force", description="布局类型: force/hierarchical/circular")


class VisualizationNode(BaseModel):
    """可视化节点Schema"""
    id: int
    label: str
    type: str
    group: str
    value: int
    title: str


class VisualizationEdge(BaseModel):
    """可视化边Schema"""
    from: int
    to: int
    label: str
    value: float
    title: str


class VisualizationResponse(BaseModel):
    """可视化响应Schema"""
    nodes: List[VisualizationNode]
    edges: List[VisualizationEdge]
    layout: str


# ============================================
# 统计相关Schema
# ============================================

class GraphStatsResponse(BaseModel):
    """图谱统计响应Schema"""
    total_entities: int
    total_relationships: int
    entity_type_distribution: Dict[str, int]
    relation_type_distribution: Dict[str, int]
    graph_density: float
    average_degree: float
    largest_component_size: int


# ============================================
# 批量操作Schema
# ============================================

class EntityBatchCreate(BaseModel):
    """批量创建实体Schema"""
    entities: List[EntityCreate]


class EntityBatchUpdate(BaseModel):
    """批量更新实体Schema"""
    entities: List[Dict[str, Any]]  # 包含id和更新字段


class EntityBatchDelete(BaseModel):
    """批量删除实体Schema"""
    entity_ids: List[int]


class EntityMergeRequest(BaseModel):
    """实体合并请求Schema"""
    entity_ids: List[int] = Field(..., description="要合并的实体ID列表")
    target_entity_id: int = Field(..., description="目标实体ID")

File: multikb-knowledge-frontend/src/api/modules/knowledge-graph.ts
import request from '../utils/request'
import type { PaginationResult, PaginationParams } from '@/types'

// ============================================
// 类型定义
// ============================================

export interface KnowledgeGraphEntity {
  id: number
  knowledge_base_id: number
  user_id: number
  name: string
  type: string
  description?: string
  aliases?: string[]
  confidence: number
  metadata?: Record<string, any>
  created_at: string
  updated_at: string
  related_documents_count?: number
  related_entities_count?: number
}

export interface KnowledgeGraphRelationship {
  id: number
  knowledge_base_id: number
  user_id: number
  source_entity_id: number
  target_entity_id: number
  relation_type: string
  description?: string
  weight: number
  confidence: number
  metadata?: Record<string, any>
  created_at: string
  updated_at: string
  source_entity?: KnowledgeGraphEntity
  target_entity?: KnowledgeGraphEntity
}

export interface EntityCreate {
  knowledge_base_id: number
  name: string
  type: string
  description?: string
  aliases?: string[]
  confidence?: number
  metadata?: Record<string, any>
}

export interface EntityUpdate {
  name?: string
  type?: string
  description?: string
  aliases?: string[]
  confidence?: number
  metadata?: Record<string, any>
}

export interface RelationshipCreate {
  knowledge_base_id: number
  source_entity_id: number
  target_entity_id: number
  relation_type: string
  description?: string
  weight?: number
  confidence?: number
  metadata?: Record<string, any>
}

export interface RelationshipUpdate {
  relation_type?: string
  description?: string
  weight?: number
  confidence?: number
  metadata?: Record<string, any>
}

export interface PathQuery {
  source_entity_id: number
  target_entity_id: number
  max_hops?: number
  relation_types?: string[]
}

export interface VisualizationData {
  nodes: Array<{
    id: number
    label: string
    type: string
    group: string
    value: number
    title: string
  }>
  edges: Array<{
    from: number
    to: number
    label: string
    value: number
    title: string
  }>
  layout: string
}

export interface GraphStats {
  total_entities: number
  total_relationships: number
  entity_type_distribution: Record<string, number>
  relation_type_distribution: Record<string, number>
  graph_density: number
  average_degree: number
  largest_component_size: number
}

// ============================================
// 实体相关API
// ============================================

// 获取实体列表
export const getEntities = (params: PaginationParams & {
  knowledge_base_id: number
  type?: string
  keyword?: string
}) => {
  return request<PaginationResult<KnowledgeGraphEntity>>({
    url: '/knowledge-graph/entities',
    method: 'get',
    params
  })
}

// 创建实体
export const createEntity = (data: EntityCreate) => {
  return request<KnowledgeGraphEntity>({
    url: '/knowledge-graph/entities',
    method: 'post',
    data
  })
}

// 获取实体详情
export const getEntityDetail = (id: number) => {
  return request<{
    code: number
    message: string
    data: {
      entity: KnowledgeGraphEntity
      relationships: KnowledgeGraphRelationship[]
      related_documents: Array<{
        document_id: number
        title: string
        mentions_count: number
      }>
    }
  }>({
    url: `/knowledge-graph/entities/${id}`,
    method: 'get'
  })
}

// 更新实体
export const updateEntity = (id: number, data: EntityUpdate) => {
  return request<KnowledgeGraphEntity>({
    url: `/knowledge-graph/entities/${id}`,
    method: 'put',
    data
  })
}

// 删除实体
export const deleteEntity = (id: number) => {
  return request({
    url: `/knowledge-graph/entities/${id}`,
    method: 'delete'
  })
}

// 批量创建实体
export const batchCreateEntities = (data: { entities: EntityCreate[] }) => {
  return request<{
    code: number
    message: string
    data: {
      created: number
      failed: number
      entities: KnowledgeGraphEntity[]
    }
  }>({
    url: '/knowledge-graph/entities/batch',
    method: 'post',
    data
  })
}

// 批量更新实体
export const batchUpdateEntities = (data: { entities: Array<{ id: number } & EntityUpdate> }) => {
  return request({
    url: '/knowledge-graph/entities/batch',
    method: 'put',
    data
  })
}

// 批量删除实体
export const batchDeleteEntities = (data: { entity_ids: number[] }) => {
  return request({
    url: '/knowledge-graph/entities/batch',
    method: 'delete',
    data
  })
}

// 合并实体
export const mergeEntities = (data: {
  entity_ids: number[]
  target_entity_id: number
}) => {
  return request<KnowledgeGraphEntity>({
    url: '/knowledge-graph/entities/merge',
    method: 'post',
    data
  })
}

// 搜索实体
export const searchEntities = (params: {
  q: string
  knowledge_base_id?: number
  type?: string
}) => {
  return request<{
    code: number
    message: string
    data: {
      entities: Array<KnowledgeGraphEntity & { score: number; matched_field: string }>
    }
  }>({
    url: '/knowledge-graph/search',
    method: 'get',
    params
  })
}

// ============================================
// 关系相关API
// ============================================

// 获取关系列表
export const getRelationships = (params: PaginationParams & {
  knowledge_base_id: number
  source_entity_id?: number
  target_entity_id?: number
  relation_type?: string
}) => {
  return request<PaginationResult<KnowledgeGraphRelationship>>({
    url: '/knowledge-graph/relationships',
    method: 'get',
    params
  })
}

// 创建关系
export const createRelationship = (data: RelationshipCreate) => {
  return request<KnowledgeGraphRelationship>({
    url: '/knowledge-graph/relationships',
    method: 'post',
    data
  })
}

// 获取关系详情
export const getRelationshipDetail = (id: number) => {
  return request<{
    code: number
    message: string
    data: KnowledgeGraphRelationship
  }>({
    url: `/knowledge-graph/relationships/${id}`,
    method: 'get'
  })
}

// 更新关系
export const updateRelationship = (id: number, data: RelationshipUpdate) => {
  return request<KnowledgeGraphRelationship>({
    url: `/knowledge-graph/relationships/${id}`,
    method: 'put',
    data
  })
}

// 删除关系
export const deleteRelationship = (id: number) => {
  return request({
    url: `/knowledge-graph/relationships/${id}`,
    method: 'delete'
  })
}

// ============================================
// 查询相关API
// ============================================

// 路径查询
export const findPaths = (data: PathQuery) => {
  return request<{
    code: number
    message: string
    data: {
      paths: Array<{
        entities: number[]
        relationships: Array<{
          source: number
          target: number
          type: string
        }>
        total_weight: number
        path_length: number
      }>
    }
  }>({
    url: '/knowledge-graph/paths',
    method: 'post',
    data
  })
}

// 获取实体邻居
export const getEntityNeighbors = (entityId: number, params?: {
  relation_type?: string
  max_depth?: number
  limit?: number
}) => {
  return request<{
    code: number
    message: string
    data: {
      neighbors: Array<{
        entity: KnowledgeGraphEntity
        relationship: KnowledgeGraphRelationship
        depth: number
      }>
    }
  }>({
    url: `/knowledge-graph/entities/${entityId}/neighbors`,
    method: 'get',
    params
  })
}

// 获取可视化数据
export const getVisualizationData = (params: {
  knowledge_base_id: number
  entity_ids?: number[]
  relation_types?: string[]
  entity_type?: string
  max_nodes?: number
  layout?: string
}) => {
  return request<{
    code: number
    message: string
    data: VisualizationData
  }>({
    url: '/knowledge-graph/visualization',
    method: 'get',
    params
  })
}

// ============================================
// 提取相关API
// ============================================

// 批量提取实体
export const extractEntities = (data: {
  knowledge_base_id: number
  document_ids?: number[]
  entity_type_mode?: 'system' | 'user' | 'model'
  entity_type_codes?: string[]
}) => {
  return request<{
    code: number
    message: string
    data: {
      task_id: number
      status: string
    }
  }>({
    url: '/knowledge-graph/extract',
    method: 'post',
    data
  })
}

// 单文档提取
export const extractDocumentEntities = (documentId: number, entity_type_mode?: 'system' | 'user' | 'model') => {
  return request<{
    code: number
    message: string
    data: {
      task_id: number
      status: string
    }
  }>({
    url: `/knowledge-graph/extract/document/${documentId}`,
    method: 'post',
    params: entity_type_mode ? { entity_type_mode } : undefined
  })
}

// 提取任务相关类型
export interface ExtractionTask {
  id: number
  knowledge_base_id: number
  document_id?: number
  status: 'pending' | 'processing' | 'completed' | 'failed'
  total_entities: number
  total_relationships: number
  error_message?: string
  metadata?: Record<string, any>
  created_at: string
  updated_at: string
  completed_at?: string
}

// 获取提取任务列表
export const listExtractionTasks = (params: { knowledge_base_id: number; page?: number; size?: number }) => {
  return request<{
    code: number
    message: string
    data: {
      tasks: ExtractionTask[]
      total: number
      page: number
      size: number
    }
  }>({
    url: '/knowledge-graph/extract/tasks',
    method: 'get',
    params
  })
}

// 获取提取任务状态
export const getExtractionTask = (taskId: number) => {
  return request<{
    code: number
    message: string
    data: ExtractionTask
  }>({
    url: `/knowledge-graph/extract/tasks/${taskId}`,
    method: 'get'
  })
}

// 删除提取任务
export const deleteExtractionTask = (taskId: number) => {
  return request<{
    code: number
    message: string
  }>({
    url: `/knowledge-graph/extract/tasks/${taskId}`,
    method: 'delete'
  })
}

// 重新生成提取任务
export const regenerateExtractionTask = (taskId: number) => {
  return request<{
    code: number
    message: string
    data: {
      old_task_id: number
      new_task_id: string
      document_id: number
    }
  }>({
    url: `/knowledge-graph/extract/tasks/${taskId}/regenerate`,
    method: 'post'
  })
}

// ============================================
// 实体类型管理API
// ============================================

export interface EntityType {
  id: number
  code: string
  name: string
  description?: string
  icon?: string
  color?: string
  tag_type?: string
  sort_order: number
  is_system: boolean
  is_enabled: boolean
  parent_id?: number | null
  level: number
  children?: EntityType[]
  metadata?: Record<string, any>
  created_at: string
  updated_at: string
}

export interface EntityTypeCreate {
  code: string
  name: string
  description?: string
  icon?: string
  color?: string
  tag_type?: string
  sort_order?: number
  is_enabled?: boolean
  parent_id?: number | null
  level?: number
  metadata?: Record<string, any>
}

export interface EntityTypeUpdate {
  name?: string
  description?: string
  icon?: string
  color?: string
  tag_type?: string
  sort_order?: number
  is_enabled?: boolean
  parent_id?: number | null
  level?: number
  metadata?: Record<string, any>
}

// 获取实体类型列表
export const getEntityTypes = (params?: {
  is_enabled?: boolean
  is_system?: boolean
  knowledge_base_id?: number
  level?: number
  parent_id?: number
  include_children?: boolean
  page?: number
  size?: number
  keyword?: string
}) => {
  return request<{
    code: number
    message: string
    data: {
      list?: EntityType[]
      items?: EntityType[]
      types?: EntityType[]  // 兼容旧格式
      total: number
      page?: number
      size?: number
      pages?: number
      has_next?: boolean
      has_prev?: boolean
    }
  }>({
    url: '/knowledge-graph/entity-types',
    method: 'get',
    params
  })
}

// 获取实体类型树
export const getEntityTypeTree = (params?: {
  root_id?: number
  knowledge_base_id?: number
}) => {
  return request<{
    code: number
    message: string
    data: {
      tree: EntityType[]
      total: number
    }
  }>({
    url: '/knowledge-graph/entity-types/tree',
    method: 'get',
    params
  })
}

// 获取指定类型的子类型
export const getEntityTypeChildren = (typeId: number) => {
  return request<{
    code: number
    message: string
    data: {
      types: EntityType[]
      total: number
    }
  }>({
    url: `/knowledge-graph/entity-types/${typeId}/children`,
    method: 'get'
  })
}

// 获取实体类型详情
export const getEntityType = (typeId: number) => {
  return request<{
    code: number
    message: string
    data: EntityType
  }>({
    url: `/knowledge-graph/entity-types/${typeId}`,
    method: 'get'
  })
}

// 创建实体类型
export const createEntityType = (data: EntityTypeCreate) => {
  return request<{
    code: number
    message: string
    data: EntityType
  }>({
    url: '/knowledge-graph/entity-types',
    method: 'post',
    data
  })
}

// 更新实体类型
export const updateEntityType = (typeId: number, data: EntityTypeUpdate) => {
  return request<{
    code: number
    message: string
    data: EntityType
  }>({
    url: `/knowledge-graph/entity-types/${typeId}`,
    method: 'put',
    data
  })
}

// 删除实体类型
export const deleteEntityType = (typeId: number) => {
  return request<{
    code: number
    message: string
  }>({
    url: `/knowledge-graph/entity-types/${typeId}`,
    method: 'delete'
  })
}

// 启用/禁用实体类型
export const toggleEntityType = (typeId: number, isEnabled: boolean) => {
  return request<{
    code: number
    message: string
    data: EntityType
  }>({
    url: `/knowledge-graph/entity-types/${typeId}/toggle`,
    method: 'patch',
    params: { is_enabled: isEnabled }
  })
}

// 获取知识库的实体类型配置
export const getKnowledgeBaseEntityTypes = (kbId: number) => {
  return request<{
    code: number
    message: string
    data: {
      types: EntityType[]
      total: number
    }
  }>({
    url: `/knowledge-graph/knowledge-bases/${kbId}/entity-types`,
    method: 'get'
  })
}

// 配置知识库的实体类型
export const configureKnowledgeBaseEntityTypes = (
  kbId: number,
  data: {
    entity_type_ids: number[]
    configs?: Array<{
      entity_type_id: number
      is_enabled: boolean
      sort_order: number
    }>
  }
) => {
  return request<{
    code: number
    message: string
    data: {
      count: number
    }
  }>({
    url: `/knowledge-graph/knowledge-bases/${kbId}/entity-types`,
    method: 'post',
    data
  })
}

// ============================================
// 行业模板API
// ============================================

export interface IndustryTemplate {
  code: string
  name: string
  description: string
  entity_type_count: number
}

export interface IndustryTemplateDetail {
  template: {
    code: string
    name: string
    description: string
  }
  entity_types: Array<EntityType & { sort_order?: number; exists?: boolean }>
}

// 获取所有行业模板列表
export const getIndustryTemplates = () => {
  return request<{
    code: number
    message: string
    data: {
      templates: IndustryTemplate[]
      total: number
    }
  }>({
    url: '/knowledge-graph/entity-type-templates',
    method: 'get'
  })
}

// 获取指定行业模板详情
export const getIndustryTemplate = (templateCode: string) => {
  return request<{
    code: number
    message: string
    data: IndustryTemplateDetail
  }>({
    url: `/knowledge-graph/entity-type-templates/${templateCode}`,
    method: 'get'
  })
}

// 应用行业模板到知识库
export const applyIndustryTemplate = (kbId: number, templateCode: string) => {
  return request<{
    code: number
    message: string
    data: {
      count: number
    }
  }>({
    url: `/knowledge-graph/knowledge-bases/${kbId}/apply-template`,
    method: 'post',
    data: {
      template_code: templateCode
    }
  })
}

// ============================================
// 统计相关API
// ============================================

// 获取统计信息
export const getGraphStats = (knowledgeBaseId: number) => {
  return request<{
    code: number
    message: string
    data: GraphStats
  }>({
    url: '/knowledge-graph/stats',
    method: 'get',
    params: { knowledge_base_id: knowledgeBaseId }
  })
}

